File: desktop/macos/Sources/NexusDesktop/Sidecar/SidecarPortAllocator.swift
import Darwin
import Foundation

enum SidecarPortAllocator {
  private static let desktopLoopbackPort = 34343
  private static let portRetryCount = 100
  private static let portRetryIntervalMicroseconds: useconds_t = 200_000

  static func allocate(startupTimeline: DesktopStartupTimeline? = nil) throws -> Int {
    if isAvailable(desktopLoopbackPort) {
      return desktopLoopbackPort
    }

    startupTimeline?.mark("sidecar.port_wait_begin", metadata: [
      "port": "\(desktopLoopbackPort)",
    ])
    // 旧 sidecar 正在响应退出信号时，给固定本地端口一个短暂释放窗口。
    for attempt in 1...portRetryCount {
      usleep(portRetryIntervalMicroseconds)
      if isAvailable(desktopLoopbackPort) {
        startupTimeline?.mark("sidecar.port_wait_ready", metadata: [
          "attempt": "\(attempt)",
          "port": "\(desktopLoopbackPort)",
        ])
        return desktopLoopbackPort
      }
    }

    startupTimeline?.mark("sidecar.port_unavailable", metadata: [
      "port": "\(desktopLoopbackPort)",
    ])
    throw DesktopShellError.portUnavailable(desktopLoopbackPort)
  }

  private static func isAvailable(_ port: Int) -> Bool {
    let socketFD = socket(AF_INET, SOCK_STREAM, 0)
    if socketFD < 0 {
      return false
    }
    defer {
      close(socketFD)
    }

    // macOS 刚关闭的 HTTP 连接可能让不带 SO_REUSEADDR 的探测 socket 误判端口仍被占用。
    var reuseAddress: Int32 = 1
    _ = setsockopt(
      socketFD,
      SOL_SOCKET,
      SO_REUSEADDR,
      &reuseAddress,
      socklen_t(MemoryLayout<Int32>.size)
    )

    var address = sockaddr_in()
    address.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
    address.sin_family = sa_family_t(AF_INET)
    address.sin_port = in_port_t(port).bigEndian
    address.sin_addr = in_addr(s_addr: inet_addr("127.0.0.1"))

    let result = withUnsafePointer(to: &address) { pointer in
      pointer.withMemoryRebound(to: sockaddr.self, capacity: 1) { socketAddress in
        bind(socketFD, socketAddress, socklen_t(MemoryLayout<sockaddr_in>.size))
      }
    }
    return result == 0
  }
}

File: desktop/windows/Nexus.Desktop/Sidecar/SidecarPortAllocator.cs
using System.Net;
using System.Net.Sockets;
using Nexus.Desktop.Diagnostics;

namespace Nexus.Desktop.Sidecar;

internal static class SidecarPortAllocator
{
    private const int DesktopLoopbackPort = 34343;
    private const int PortRetryCount = 100;
    private const int PortRetryIntervalMilliseconds = 200;

    public static int Allocate(DesktopStartupTimeline? startupTimeline = null)
    {
        if (IsAvailable(DesktopLoopbackPort))
        {
            return DesktopLoopbackPort;
        }

        startupTimeline?.Mark("sidecar.port_wait_begin", new Dictionary<string, string>
        {
            ["port"] = DesktopLoopbackPort.ToString(),
        });
        // 旧 sidecar 正在响应退出信号时，给固定本地端口一个短暂释放窗口。
        for (int attempt = 1; attempt <= PortRetryCount; attempt++)
        {
            Thread.Sleep(PortRetryIntervalMilliseconds);
            if (IsAvailable(DesktopLoopbackPort))
            {
                startupTimeline?.Mark("sidecar.port_wait_ready", new Dictionary<string, string>
                {
                    ["attempt"] = attempt.ToString(),
                    ["port"] = DesktopLoopbackPort.ToString(),
                });
                return DesktopLoopbackPort;
            }
        }

        startupTimeline?.Mark("sidecar.port_unavailable", new Dictionary<string, string>
        {
            ["port"] = DesktopLoopbackPort.ToString(),
        });
        throw new InvalidOperationException($"Nexus 桌面端本地端口 {DesktopLoopbackPort} 已被占用，请关闭占用该端口的进程后重试。");
    }

    private static bool IsAvailable(int port)
    {
        try
        {
            using var listener = new TcpListener(IPAddress.Loopback, port);
            listener.Start();
            return true;
        }
        catch (SocketException)
        {
            return false;
        }
    }
}
